Support for midori

alfonsosanchezbeato opened the pull request.

tonyespy commented on drivers/mtk2modem/gprs.c:
```diff
@@ -0,0 +1,93 @@
+#include <stdlib.h>
+#include <stdio.h>
+#include <errno.h>
+
```
Are all of the above includes necessary?  errno.h is needed, but I don't see any functions/constants used from the other headers.

tonyespy commented on drivers/mtk2modem/gprs.c:
```diff
@@ -0,0 +1,93 @@
+#include <ofono/modem.h>
+#include <ofono/gprs.h>
+#include <ofono/types.h>
+
```
Same question for the ofono headers...

alfonsosanchezbeato replied:
The perils of copy and paste. Reduced to the minimum set now.

tonyespy commented on drivers/mtk2modem/mtk2modem.c:
```diff
@@ -0,0 +1,55 @@
+}
+
+OFONO_PLUGIN_DEFINE(mtk2modem, "MTK v2 modem driver", VERSION,
+		OFONO_PLUGIN_PRIORITY_DEFAULT, mtk2modem_init, mtk2modem_exit)
```
Is there any other identifier we could use besides "v2" for this modem?  It's not really "version 2", it's a new modem from MTK.  I think I'm still OK with using "2" in the driver name, but having some description which actually binds this driver to an actual hw device would be useful.

alfonsosanchezbeato replied:
Description changed to "MTK2 modem driver". However, I prefer to not explicitly mention midori here, as (hopefully) all future MTK modems will use this new driver (or mtk3, even).

tonyespy commented on drivers/rilmodem/call-barring.c:
```diff
@@ -33,6 +33,7 @@
 #include "call-barring.h"
 #include "rilmodem.h"
 #include "ril_constants.h"
+#include "rilutil.h"
```
Why have you added "rilutil.h" to all of these driver atoms?

alfonsosanchezbeato replied:
I removed it from rilmodem.h and moved to the atoms because the definitions are not actually used in rilmodem.h and because rilmodem.h is included in ril plugins files (like mtk2.c) that do not need rilutil.h too.

tonyespy commented on drivers/rilmodem/gprs.c:
```diff
@@ -128,7 +134,12 @@ void ril_gprs_set_ia_apn(struct ofono_gprs *gprs, const char *apn,
+	/*
+	 * FIXME for the moment we use ril version to avoid the need to set
+	 * the env var OFONO_RIL_RAT_LTE in latest modems.
+	 */
```
Perhaps it might be worth a comment stating the RIL v11 first appeared in Marshmallow?

alfonsosanchezbeato replied:
Comment added in ril.c, when g_ril_set_version() is called. This part of the code is removed in the new version that I will push.

tonyespy commented on drivers/rilmodem/sim.c:
```diff
@@ -943,7 +944,8 @@ static void ril_query_pin_retries(struct ofono_sim *sim,
 		}
-	} else if (sd->vendor == OFONO_RIL_VENDOR_MTK) {
+	} else if (sd->vendor == OFONO_RIL_VENDOR_MTK ||
+			sd->vendor == OFONO_RIL_VENDOR_MTK2) {
```
Hmmm...  I don't really like this.  We're really quirking on a different modem design from the same vendor, not a new vendor.

alfonsosanchezbeato replied:
My testing has revealed that some quirk are needed for:

Only OFONO_RIL_VENDOR_MTK (like in g_ril_reply_parse_avail_ops())
Only OFONO_RIL_VENDOR_MTK2 (like in g_ril_unsol_parse_connected())
Both (most other quirks)

So we need both vendor tags.

tonyespy replied:
I understand the need for the different quirks, my point was that you're using a "vendor" field for the quirking when the *vendor* hasn't changed.  These are really model ( or family ) specific quirks.  If MTK does yet another re-design, we'd end up with MTK3, which in my mind is less than ideal.  Mind adding a note to the effect that this should be re-factored if/when yet another design lands?

alfonsosanchezbeato replied:
More than refactoring, maybe this should be called model/family/type or something like that instead of "vendor". Note also that in "enum ofono_vendor", defined by the atmodem driver we have:

	OFONO_VENDOR_GOBI,    (which is a qualcomm modem)
	OFONO_VENDOR_QUALCOMM_MSM,
	OFONO_VENDOR_WAVECOM,
	OFONO_VENDOR_WAVECOM_Q2XXX,
	OFONO_VENDOR_SIMCOM,
	OFONO_VENDOR_SIMCOM_SIM900,

so if changed, we should do that for this enum too.

tonyespy commented on drivers/rilmodem/ussd.c:
```diff
@@ -178,8 +180,10 @@ static void ril_ussd_notify(struct ril_msg *message, gpointer user_data)
+	vendor = g_ril_vendor(ud->ril);
+	if ((vendor == OFONO_RIL_VENDOR_MTK ||
+				vendor == OFONO_RIL_VENDOR_MTK2)
+			&& unsol->message != NULL && unsol->type == 2)
```
Same comment as above...

tonyespy commented on gril/ril_constants.h:
```diff
@@ -427,4 +465,37 @@
 
+/* Radio capability phases */
+#define RIL_RC_PHASE 0
+#define RIL_RC_PHASE_CONFIGURED 1
```
These enums seem to be off by 1 compared the values in AOSP v6 version of ril.h ( version 11 ).

alfonsosanchezbeato replied:
Good catch. Changed now.

tonyespy commented on plugins/ril.c:
```diff
@@ -77,83 +81,162 @@ struct ril_data {
+			DBG("calling set_powered(TRUE)");
+			rd->init_state = FALSE;
+
+			/* Note that hw is powered, but offline */
```
Not sure what you mean by this comment...

alfonsosanchezbeato replied:
Changed to "Note that modem hw is powered, but radio is off"

tonyespy commented on plugins/ril.c:
```diff
@@ -77,83 +81,162 @@ struct ril_data {
+		 * the event.
+		 */
+		if (rd->ofono_online)
+			ofono_info("%s: radio self-powered off!",
```
This is a major change in semantics...  So what happens if the radio powers off and just stays like that?

alfonsosanchezbeato replied:
Unfortunately we have the issue mentioned in the comment when entering SIM PIN. Besides, tbh I have never seen the modem auto-powering off, that would be a bug in ril/modem fw and I am not sure on how can one actually protect from that.

Also, with the new logic in place, where we have an init state where we make sure radio state is OFF before doing anything, I have not seen STATE_UNAVAILABLE happening unexpectedly on start, not even in mako. This fixes the issue with ofonod restarting some times on boot for both turbo and mako.

Note also that the assertion and call to exit were removed in upstream version.

Finally, I have some additional changes in this logic that I will push now. The observation that led to these changes is that the actual radio state does not actually change until you get the radio state change event. Just waiting for the reply to RADIO request provoked errors when that happened before the event.

tonyespy commented on plugins/ril.c:
```diff
@@ -267,19 +355,27 @@ void ril_post_sim(struct ofono_modem *modem)
+			get_driver_type(rd, OFONO_ATOM_TYPE_CALL_FORWARDING),
+			rd->ril);
+	ofono_phonebook_create(modem, rd->vendor,
+			get_driver_type(rd, OFONO_ATOM_TYPE_PHONEBOOK), modem);
```
Any reason why the ordering of these calls was reversed?

alfonsosanchezbeato replied:
Not sure now why I did this. However, in new version I have moved ofono_call_forwarding_create to post_online, in the same place as all SS services are created (call barring, call settings, etc.).

tonyespy commented on src/gprs.c:
```diff
@@ -1984,9 +1984,10 @@ static DBusMessage *gprs_set_property(DBusConnection *conn,
+
 		gprs_netreg_update(gprs);
 
-		notify_powered_change(gprs);
```
Why the ordering change here?

alfonsosanchezbeato replied:
notify_powered_change() does things only when we power on data for a modem. Changing the order here makes sure that we first detach the other modems before trying to attach for the modem, which makes more sense than the other way around (we are "freeing" the radio resources before trying to attach from a different slot).

tonyespy replied:
Please submit this change upstream.

alfonsosanchezbeato replied:
The changes to add notify_powered_change() were not submitted. It was not clear how to implement dual SIM from upstream's POV. Although I agree on start submitting things as soon as we rebase to ofono 1.18.

tonyespy commented on plugins/ril.h:
```diff
@@ -28,3 +33,5 @@ void ril_post_sim(struct ofono_modem *modem);
 void ril_set_online(struct ofono_modem *modem, ofono_bool_t online,
 			ofono_modem_online_cb_t callback, void *data);
+
+GRil *ril_get_gril_c(struct ofono_modem *modem);
```
In plugins/mtk.c you use the full word "complement" vs. "c".  That might be a more obvious choice for this function name too.

alfonsosanchezbeato replied:
Changed.

tonyespy commented on gril/grilreply.c:
```diff
@@ -1456,3 +1461,99 @@ struct parcel_str_array *g_ril_reply_oem_hook_strings(GRil *gril,
 	return str_arr;
 }
+
+struct reply_radio_capability *g_ril_reply_parse_get_radio_capability(
```
Considering the fact that upstream nak'd this whole approach, perhaps it would be better to not continue to add new functions to grilreply/request/unsol, and move this code inline where it's used?

Also, no unit tests exist for this ( or for any of the other message changes ).

alfonsosanchezbeato replied:
In the new version I have used this function in two places: radio-settings driver and ril plugin, so it is useful to have this in grilreply. Let's leave this here for the moment and refactor everything when we update to ofono 1.18 (which has been recently released).

tonyespy replied:
OK, I'll let this slide for now, but any future code should be in-lined...

alfonsosanchezbeato replied:
I will get this inlined in the next PR that will support switching the 4G slot.

tonyespy commented:
Overall this looks good, although there are some major changes, which will require a size-able testing effort.

What devices have you sanity tested so far ( ie. ensure that they power, and that data works )?

Also, based upon the changes, it would seem that Midori is a dual-SIM phone?

alfonsosanchezbeato commented:
I have tested on all ril-based devices: mako, turbo (solving the issue we had there with switching data between SIMs), and midori. Basic stuff works well. This change needs indeed a big testing effort in all fronts.

Yes, midori is a dual-SIM phone. The nice thing with it and turbo is that they seem to try to use standard AOSP multi-sim support instead of their own thing.

alfonsosanchezbeato commented:
Branch refreshed after addressing comments. Some additional changes performed in ril and rildev.

tonyespy commented:
Overall, the changes look good however:

1. No unit tests were added to cover the new requests and/or MTK2 quirks.  I'm not going to hold up the merge, but please add a card to our trello board to resolve.

2. There's still no comment that describes the mapping of actual hardware to MTK2 driver.  Pretty sure we haven't upstreamed any of our MTK code yet, however I imagine this will be an issue when we do.  Can you add an Asana task to investigate whether or not we can identify these modems using udev, or some other method ( vs. our current approach of specifying an env var that's used to load the proper driver )?

3. Can you explain why you dropped the assert on malformed radio messages?  Was this something you actually encountered with Midori?

4. You also changed the code to not exit when the radio asynchronously powers off.  The reason this was done was that the modem could be left in an inconsistent state. If the new MTK modem can generate such an unsolicited event, then perhaps you should use a quirk to handle this.

5. The commit: "ril: create gprs atom after onlining" actually does a lot more:

 * call_forwarding and radio_settings also moved to post_online
 * set_online / send_power logic re-factored
 * assert on malformed radio power messages removed
 * code no longer exits when asynchronous power off received

I'm not going to ask that you split all the code out into separate commits ( which is what should've be done ), but I will ask that you amend the commit message and mention the other changes.

alfonsosanchezbeato commented:
Thanks for the review. Branch refreshed after addressing comments. Answering the bullets from the final comment:

1. Task added to add tests: https://app.asana.com/0/39075168339873/149012010444678

2. It is difficult to know the SoCs where the MTK2 driver will be needed, as we have no information from the vendor. Best guess it that starting from midori's SoC all following MTK modems will need it. In any case, removing the env vars by providing better detection has been always on my list. Task added too: https://app.asana.com/0/39075168339873/149012010444681 .

3. The assert was removed by upstream too. Also, I do not remember having seen this in any modem (MTKv1 modems have more states, but that's handled by their plugin). I do not think we should exit if that happens anyway, and upstream is of the same opinion. Note that the ofono_info() in the function will let us know if this happens by looking at syslog.

4. Yes, there is a glitch in midori when inserting SIM PIN: radio transitions from ON->OFF->ON again very quickly. I have added back a comment on this (I had this but seems I dropped it without noticing in one of the refactors I did). Re: removing the call to exit(), I did that for 3 reasons: first upstream also removed it, second I have never been sure the correct thing to do was to exit here, third with the refactoring we are not receiving unexpected events on boot any more for the devices that showed them (mako and turbo).

5. Done

tonyespy commented:
LGTM
